pkg/gardenlet/controller/bastion/reconciler.go

```diff
@@ -130,54 +123,90 @@ func (r *reconciler) reconcileBastion(
-		metav1.SetMetaDataAnnotation(&extBastion.ObjectMeta, v1beta1constants.GardenerTimestamp, time.Now().UTC().String())
-		extBastion.Spec.UserData = createUserData(bastion)
-		extBastion.Spec.Ingress = extensionsIngress
-		extBastion.Spec.Type = *bastion.Spec.ProviderType
-		return nil
-	}); err != nil {
-		if patchErr := patchReadyCondition(ctx, r.gardenClient, bastion, gardencorev1alpha1.ConditionFalse, "FailedReconciling", err.Error()); patchErr != nil {
-			log.Error(patchErr, "Failed patching ready condition")
+	var (
+		mustReconcileExtensionBastion = false
+		reconciliationSuccessful      = true
+		lastObservedError             error
+		extensionBastionSpec          = extensionsv1alpha1.BastionSpec{
+			DefaultSpec: extensionsv1alpha1.DefaultSpec{
+				Type: *bastion.Spec.ProviderType,
+			},
+			UserData: createUserData(bastion),
+			Ingress:  extensionsIngress,
 		}
+	)
 
-		return fmt.Errorf("failed to ensure bastion extension resource: %w", err)
+	if err := r.SeedClient.Get(ctx, client.ObjectKeyFromObject(extBastion), extBastion); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		// if the extension Bastion doesn't exist yet, create it
+		mustReconcileExtensionBastion = true
+	} else if !reflect.DeepEqual(extBastion.Spec, extensionBastionSpec) {
+		// if the extensionBastionSpec has changed, reconcile it
+		mustReconcileExtensionBastion = true
+	} else if extBastion.Status.LastOperation == nil {
+		reconciliationSuccessful = false
+		lastObservedError = fmt.Errorf("extension did not record a last operation yet")
+	} else {
+		lastOperationState := extBastion.Status.LastOperation.State
+		if extBastion.Status.LastError != nil ||
+			lastOperationState == gardencorev1beta1.LastOperationStateError ||
+			lastOperationState == gardencorev1beta1.LastOperationStateFailed {
+			if lastOperationState == gardencorev1beta1.LastOperationStateFailed {
+				mustReconcileExtensionBastion = true
+			}
+			reconciliationSuccessful = false
+
+			lastObservedError = fmt.Errorf("extension state is not Succeeded but %v", lastOperationState)
+			if extBastion.Status.LastError != nil {
+				lastObservedError = v1beta1helper.NewErrorWithCodes(fmt.Errorf("error during reconciliation: %s", extBastion.Status.LastError.Description), extBastion.Status.LastError.Codes...)
+			}
+		}
 	}
 
-	// wait for the extension controller to reconcile possible changes
-	if err := extensions.WaitUntilExtensionObjectReady(
-		ctx,
-		r.seedClient,
-		log,
-		extBastion,
-		extensionsv1alpha1.BastionResource,
-		defaultInterval,
-		defaultSevereThreshold,
-		defaultTimeout,
-		nil,
-	); err != nil {
-		if patchErr := patchReadyCondition(ctx, r.gardenClient, bastion, gardencorev1alpha1.ConditionFalse, "FailedReconciling", err.Error()); patchErr != nil {
+	if lastObservedError != nil {
+		message := fmt.Sprintf("Error while waiting for %s to become ready", extensionKey(extensionsv1alpha1.BastionResource, extBastion.Namespace, extBastion.Name))
```

Comment:
I don't think we need this `extensionKey` function - it is only used here and you can just format it here as needed.